snippets/firestore/firestore_pipelines.py

```diff
@@ -0,0 +1,1289 @@
+)
+from google.cloud.firestore_v1.base_vector_query import DistanceMeasure
+from google.cloud.firestore_v1.vector import Vector
+from google.cloud.firestore_v1.client import Client
+
+import firebase_admin
+from firebase_admin import firestore
+
+default_app = firebase_admin.initialize_app()
+client = firestore.client(default_app, "your-new-enterprise-database")
+
+# pylint: disable=invalid-name
+def pipeline_concepts():
+    # [START pipeline_concepts]
+    pipeline = client.pipeline() \
+        .collection("cities") \
+        .where(Field.of("population").greater_than(100000)) \
+        .sort(Field.of("name").ascending()) \
+        .limit(10)
+    # [END pipeline_concepts]
+    print(pipeline)
+
+def basic_read():
+    # [START basic_read]
+    pipeline = client.pipeline().collection("users")
+    for result in pipeline.execute():
+        print(result.id + " => " + result.data)
+    # [END basic_read]
+
+def pipeline_initialization():
+    # [START pipeline_initialization]
+    firestore_client = firestore.client(default_app, "your-new-enterprise-database")
+    pipeline = firestore_client.pipeline()
+    # [END pipeline_initialization]
+    print(pipeline)
+
+def field_vs_constants():
+    # [START field_or_constant]
+    pipeline = client.pipeline() \
+        .collection("cities") \
+        .where(Field.of("name").equal(Constant("Toronto")))
+    # [END field_or_constant]
+    print(pipeline)
+
+def input_stages():
+    # [START input_stages]
+    # Return all restaurants in San Francisco
+    results = client.pipeline().collection("cities/sf/restaurants").execute()
+
+    # Return all restaurants
+    results = client.pipeline().collection_group("restaurants").execute()
+
+    # Return all documents across all collections in the database (the entire database)
+    results = client.pipeline().database().execute()
+
+    # Batch read of 3 documents
+    results = client.pipeline().documents(
+        client.collection("cities").document("SF"),
+        client.collection("cities").document("DC"),
+        client.collection("cities").document("NY")
+    ).execute()
+    # [END input_stages]
+    for result in results:
+        print(result)
+
+def where_pipeline():
+    # [START pipeline_where]
+    results = client.pipeline().collection("books") \
+        .where(Field.of("rating").equal(5)) \
+        .where(Field.of("published").less_than(1900)) \
+        .execute()
+
+    results = client.pipeline().collection("books") \
+        .where(And(
+            Field.of("rating").equal(5),
+            Field.of("published").less_than(1900)
+        )) \
+        .execute()
+    # [END pipeline_where]
+    for result in results:
+        print(result)
+
+def aggregate_groups():
+    # [START aggregate_groups]
+    results = client.pipeline() \
+        .collection("books") \
+        .aggregate(
+            Field.of("rating").average().as_("avg_rating"),
+            groups=[Field.of("genre")]
+        ) \
+        .execute()
```

Comment:
If you want to get rid of all the \\, IIRC think you could do something like this:

```
results = (
    client.pipeline()
    .collection("books")
    .aggregate(
        Field.of("rating").average().as_("avg_rating"),
        groups=[Field.of("genre")]
    ).execute()
)
```

(But test it first)